File: src/c47/distributions/pareto.c
// SPDX-License-Identifier: GPL-3.0-only
// SPDX-FileCopyrightText: Copyright The C47 Authors

/********************************************//**
 * \file pareto.c
 ***********************************************/

#include "c47.h"

bool_t checkParamGPD(real_t *x, real_t *mu, real_t *sigma, real_t *xi, bool_t qf) {
  real_t t;

  if(!saveLastX())
    return false;

  if(!getRegisterAsReal(REGISTER_X, x)
      || !getRegisterAsReal(REGISTER_M, mu)
      || !getRegisterAsReal(REGISTER_S, sigma)
      || !getRegisterAsReal(REGISTER_Q, xi))
    goto err;

  if(realIsNegative(x)) {
    displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
    #if (EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function checkParamGPD:", "cannot calculate for x < 0", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    goto err;
  }
  else if(realIsZero(sigma) || realIsNegative(sigma)) {
    displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
    #if (EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function checkParamGPD:", "the parameter sigma must be positive", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    goto err;
  }

  // Check support range
  if(qf || realIsZero(xi))
    return true;                                // xi = 0, full line
  realSubtract(mu, sigma, &t, &ctxtReal39);
  realDivide(&t, xi, &t, &ctxtReal39);
  if(realIsNegative(xi))
    return realCompareLessEqual(x, &t);         // xi < 0, (-inf, (mu-sigma)/xi]
  return realCompareGreaterEqual(x, &t);        // xi > 0, [(mu-sigma)/xi, +inf)

  err:
  if(getSystemFlag(FLAG_SPCRES)) {
    convertRealToResultRegister(const_NaN, REGISTER_X, amNone);
  }
  return false;
}

void fnParetoP(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

  if(checkParamGPD(&x, &mu, &sigma, &xi, false)) {
    //WP34S_Pdf_F(&val, &d1, &d2, &ans, &ctxtReal39);
    convertRealToResultRegister(&ans, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnParetoL(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

  if(checkParamGPD(&x, &mu, &sigma, &xi, false)) {
    //WP34S_Cdf_F(&val, &d1, &d2, &ans, &ctxtReal39);
    convertRealToResultRegister(&ans, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnParetoU(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

  if(checkParamGPD(&x, &mu, &sigma, &xi, false)) {
    //WP34S_Cdfu_F(&val, &d1, &d2, &ans, &ctxtReal39);
    convertRealToResultRegister(&ans, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnParetoI(uint16_t unusedButMandatoryParameter) {
  real_t x, mu, sigma, xi;

    if(checkParamGPD(&x, &mu, &sigma, &xi, true)) {
      //WP34S_Qf_F(&val, &d1, &d2, &ans, &ctxtReal39);
      if(realIsNaN(&ans)) {
        displayDomainErrorMessage(ERROR_NO_ROOT_FOUND, ERR_REGISTER_LINE, REGISTER_X);
        #if (EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fGPDI:", "WP34S_Qf_F did not converge", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
        if(getSystemFlag(FLAG_SPCRES)) {
          convertRealToResultRegister(const_NaN, REGISTER_X, amNone);
        }
        return;
      }
      convertRealToResultRegister(&ans, REGISTER_X, amNone);
      adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
    }
}

File: src/shell/distributions/pareto.zig
// SPDX-License-Identifier: GPL-3.0-only
//
// Zig owner for the Pareto distribution commands (generalised Pareto / GPD,
// Type I and Type II), porting `src/c47/distributions/pareto.c`. The eight fn*
// entry points are dispatched from the items.c function table (exported via
// frontier.zig). All decNumber/register access goes through
// frontier_distribution_runtime. The testSuite has no Pareto dispatch entry, so
// these are verified by the dedicated distributions parity harness instead.

const dr = @import("distribution_runtime.zig");
const real_t = dr.real_t;

// Shared parameter check for both GPD types. `mu` is non-null for Type II only.
fn checkParamGPD(x: *real_t, mu: ?*real_t, sigma: *real_t, alpha: *real_t, qf: bool) linksection(dr.code_section) bool {
    if (!dr.saveLastX()) {
        return false;
    }

    if (!dr.getRegisterAsReal(dr.REGISTER_X, x) or
        (mu != null and !dr.getRegisterAsReal(dr.REGISTER_M, mu.?)) or
        !dr.getRegisterAsReal(dr.REGISTER_S, sigma) or
        !dr.getRegisterAsReal(dr.REGISTER_Q, alpha))
    {
        dr.specialResultNaN();
        return false;
    }

    if (dr.realIsZero(sigma) or dr.realIsNegative(sigma)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamGPD:", "the parameter sigma must be positive", null, null);
        dr.specialResultNaN();
        return false;
    }
    if (!qf and dr.realLessThan(x, if (mu == null) sigma else mu.?)) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamGPD:", "cannot calculate for x < sigma/mu", null, null);
        dr.specialResultNaN();
        return false;
    }
    return true;
}

// Type I Pareto distribution

pub fn paretoP(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;
    var u: real_t = undefined;

    if (checkParamGPD(&x, null, &sigma, &alpha, false)) {
        dr.realDivide(&sigma, &x, &t, ctx);
        dr.realPow(&t, &alpha, &u, ctx);
        dr.realDivide(&u, &x, &t, ctx);
        dr.realMultiply(&t, &alpha, &x, ctx);
        dr.storeResult(&x);
    }
}

pub fn paretoL(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var t: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;

    if (checkParamGPD(&x, null, &sigma, &alpha, false)) {
        dr.realDivide(&sigma, &x, &t, ctx);
        dr.WP34S_Ln(&t, &sigma, ctx);
        dr.realMultiply(&alpha, &sigma, &t, ctx);
        dr.realExpM1(&t, &x, ctx);
        dr.realChangeSign(&x);
        dr.storeResult(&x);
    }
}

pub fn paretoU(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamGPD(&x, null, &sigma, &alpha, false)) {
        dr.realDivide(&sigma, &x, &t, ctx);
        dr.realPow(&t, &alpha, &x, ctx);
        dr.storeResult(&x);
    }
}

pub fn paretoI(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var t: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;

    if (checkParamGPD(&x, null, &sigma, &alpha, true)) {
        if ((dr.realIsNegative(&x) and !dr.realIsZero(&x)) or dr.realGreaterThan(&x, dr.const1())) {
            dr.setNaN(&x);
        } else {
            dr.realChangeSign(&x);
            dr.WP34S_Ln1P(&x, &t, ctx);
            dr.realDivide(&t, &alpha, &x, ctx);
            dr.realChangeSign(&x);
            dr.realExponential(&x, &t, ctx);
            dr.realMultiply(&t, &sigma, &x, ctx);
        }
        dr.storeResult(&x);
    }
}

// Type II Pareto distribution

pub fn pareto2P(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamGPD(&x, &mu, &sigma, &alpha, false)) {
        dr.realSubtract(&x, &mu, &t, ctx);
        dr.realAdd(&t, &sigma, &x, ctx); // x = x + sigma - mu
        dr.realDivide(&sigma, &x, &t, ctx);
        dr.realPow(&t, &alpha, &mu, ctx);
        dr.realDivide(&mu, &x, &t, ctx);
        dr.realMultiply(&t, &alpha, &x, ctx);
        dr.storeResult(&x);
    }
}

pub fn pareto2L(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamGPD(&x, &mu, &sigma, &alpha, false)) {
        dr.realSubtract(&x, &mu, &t, ctx);
        dr.realDivide(&t, &sigma, &x, ctx);
        dr.WP34S_Ln1P(&x, &t, ctx);
        dr.realMultiply(&alpha, &t, &sigma, ctx);
        dr.realChangeSign(&sigma);
        dr.realExpM1(&sigma, &x, ctx);
        dr.realChangeSign(&x);
        dr.storeResult(&x);
    }
}

pub fn pareto2U(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamGPD(&x, &mu, &sigma, &alpha, false)) {
        dr.realSubtract(&x, &mu, &t, ctx);
        dr.realDivide(&t, &sigma, &x, ctx);
        dr.WP34S_Ln1P(&x, &t, ctx);
        dr.realMultiply(&alpha, &t, &sigma, ctx);
        dr.realChangeSign(&sigma);
        dr.realExponential(&sigma, &x, ctx);
        dr.storeResult(&x);
    }
}

pub fn pareto2I(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var mu: real_t = undefined;
    var sigma: real_t = undefined;
    var alpha: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamGPD(&x, &mu, &sigma, &alpha, true)) {
        if ((dr.realIsNegative(&x) and !dr.realIsZero(&x)) or dr.realGreaterThan(&x, dr.const1())) {
            dr.setNaN(&x);
        } else {
            dr.realChangeSign(&x);
            dr.WP34S_Ln1P(&x, &t, ctx);
            dr.realDivide(&t, &alpha, &x, ctx);
            dr.realExponential(&x, &t, ctx); // t = (1-x)^(1/alpha)
            dr.realSubtract(&mu, &sigma, &mu, ctx);
            dr.realMultiply(&mu, &t, &x, ctx);
            dr.realAdd(&x, &sigma, &mu, ctx);
            dr.realDivide(&mu, &t, &x, ctx);
        }
        dr.storeResult(&x);
    }
}
